core/capabilities/triggers/logevent/service.go

```diff
@@ -0,0 +1,154 @@
+	Logger         logger.Logger
+	Relayer        core.Relayer
+	RelayerSet     core.RelayerSet
+	LogEventConfig Config
+}
+
+var _ capabilities.TriggerCapability = (*TriggerService)(nil)
+var _ services.Service = &TriggerService{}
+
+// Creates a new Cron Trigger Service.
+// Scheduling will commence on calling .Start()
+func NewLogEventTriggerService(p Params) *TriggerService {
+	l := logger.Named(p.Logger, "LogEventTriggerCapabilityService: ")
+
+	logEventStore := NewCapabilitiesStore[logEventTrigger, capabilities.TriggerResponse]()
+
+	s := &TriggerService{
+		lggr:           l,
+		triggers:       logEventStore,
+		relayer:        p.Relayer,
+		logEventConfig: p.LogEventConfig,
+	}
+	s.CapabilityInfo, _ = s.Info(context.Background())
+	s.Validator = capabilities.NewValidator[RequestConfig, Input, capabilities.TriggerResponse](capabilities.ValidatorArgs{Info: s.CapabilityInfo})
+	return s
+}
+
+func (s *TriggerService) Info(ctx context.Context) (capabilities.CapabilityInfo, error) {
+	return capabilities.NewCapabilityInfo(
+		s.logEventConfig.Version(ID),
+		capabilities.CapabilityTypeTrigger,
+		"A trigger that listens for specific contract log events and starts a workflow run.",
+	)
+}
+
+// Register a new trigger
+// Can register triggers before the service is actively scheduling
+func (s *TriggerService) RegisterTrigger(ctx context.Context, req capabilities.TriggerRegistrationRequest) (<-chan capabilities.TriggerResponse, error) {
+	if req.Config == nil {
+		return nil, errors.New("config is required to register a log event trigger")
+	}
+	reqConfig, err := s.ValidateConfig(req.Config)
+	if err != nil {
+		return nil, err
+	}
+	// Add log event trigger with Contract details to CapabilitiesStore
+	respCh, err := s.triggers.InsertIfNotExists(req.TriggerID, func() (*logEventTrigger, chan capabilities.TriggerResponse, error) {
+		return newLogEventTrigger(ctx, reqConfig, s.logEventConfig, s.relayer)
+	})
+	if err != nil {
+		return nil, fmt.Errorf("LogEventTrigger %v", err)
+	}
+	s.lggr.Debugw("RegisterTrigger", "triggerId", req.TriggerID)
+	return respCh, nil
+}
+
+func (s *TriggerService) UnregisterTrigger(ctx context.Context, req capabilities.TriggerRegistrationRequest) error {
+	trigger, ok := s.triggers.Read(req.TriggerID)
+	if !ok {
+		return fmt.Errorf("triggerId %s not found", req.TriggerID)
+	}
+	// Close callback channel and stop log event trigger listener
+	trigger.Stop()
+	// Remove from triggers context
+	s.triggers.Delete(req.TriggerID)
+	s.lggr.Debugw("UnregisterTrigger", "triggerId", req.TriggerID)
+	return nil
+}
+
+// Start the service.
+func (s *TriggerService) Start(ctx context.Context) error {
+	if s.relayer == nil {
+		return errors.New("service has shutdown, it must be built again to restart")
+	}
+
+	return nil
+}
+
+// Close stops the Service.
+// After this call the Service cannot be started again,
+// The service will need to be re-built to start scheduling again.
+func (s *TriggerService) Close() error {
+	return nil
```

Comment:
you'll need to keep track of the logEventTrigger instances and close any live ones when the service is closed, in turn, as alluded to by jordan you will need to use a waitgroup in combination with a stop channel to ensure all goroutines are cleaned up before stopping, this is a standard pattern used in the chainlink codebase, see for example, in core,  core/capabilities/remote/target/server.go

Reply:
Sure, I've added a waitGroup and waiting for it in Stop